src/celllistimpl.h

```diff
@@ -0,0 +1,911 @@
+#pragma once
+/**
+ * @brief Cell list class templates implementation.
+ *
+ * A minimal interface to be used in other header files is provided in "celllist.h" file.
+ *
+ * @author Richard Chudoba
+ * @date 2020-02-01
+ */
+
+#include <vector>
+#include <set>
+#include <map>
+#include <algorithm>
+#include <cassert>
+#include <Eigen/Core>
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/join.hpp>
+#include <range/v3/view/view.hpp>
+#include <range/v3/view/transform.hpp>
+#include "celllist.h"
+
+namespace Faunus {
+
+/**
+ * @namespace CellList
+ * # Cell List
+ *
+ * A cell list allow to organize members, e.g., particles, into a grid cells hence allowing fast determination of their
+ * approximate location in space. Cell lists can be used to quickly compute distances or interaction only between
+ * close members. The granuality of distance is determined by the cell size and ussualy arise from some sort of cut off
+ * distance.
+ *
+ * Each cell list is composed of a grid and a container by double inheritence. The grid (Grid) is responsible for
+ * the cell indexing and the container (Container) holds he members assigned to respective cells. The class template
+ * CellListBase is a bare-bone implementation with the the essential functionality only. It should be extended using
+ * templated mixin to add features as needed.
+ *
+ * ## Cell Grid
+ *
+ * The GridType defines basic data types, GridBase calculates the cell index from the cell coordinates and vice versa,
+ * generally assuming fixed boundaries. Respective outer classes (FixedBoundaryGrid, PeriodicBoundaryGrid) wrapps
+ * boundary conditions.
+ *
+ * The convenient template aliases GridBoundary, Grid3DBoundary, Grid3DFixed and Grid3DPeriodic are provided. A simple
+ * structure GridOffsets3D contains commonly used neighbours' offsets.
+ *
+ * ## Containers
+ *
+ * Containers holds the members and allow a simple manipulation (insertion and deletion). A dense (using a vector) and
+ * sparse (using a map) implementations are provided.
+ */
+namespace CellList {
+
+/**
+ * @namespace Grid
+ */
+namespace Grid {
+
+/**
+ * @brief Interface of a grid
+ * @tparam TGridType  a cell grid types declaration
+ */
+template <typename TGridType> struct AbstractGrid : private TGridType {
+    using GridType = TGridType;
+    using typename GridType::CellCoord;
+    using typename GridType::CellIndex;
+    using typename GridType::Point;
+    using typename GridType::SpaceAxis;
```

Comment:
I understand the logic, but got confused a few times about the "SpaceAxis". Can we find a name that makes one think in the direction of a floating point type? Also, it's difficult to imagine other types than float or double, so perhaps we can simply make a typedef? One advantage is that the IDE will be able to make better code analysis.